File: lib/swift/tabbit/Updater.swift
// Tabbit's data updater.
//
// Brings a local copy of the exported data up to date with a copy served over HTTP - a
// CDN, a bucket, a patch server - so a running program can take new data without being
// redeployed. Emitted beside the reader and reads nothing but the manifest, so it knows
// nothing about the schema and never has to change when one does.
//
// The manifest is what the exporter already writes next to the data: one entry per file
// with its size and MD5. Comparing it with the local copy is the whole of the diff, so a
// run downloads what changed and nothing else.
//
// Three properties, because a patcher that fails badly is worse than one that does not
// exist:
//
//   Nothing is replaced until everything has arrived and been checked. Files land in a
//   staging directory first and the local manifest is written last, so an update killed
//   halfway leaves the previous data readable and the next run redoes the difference.
//
//   Every file is checked against the hash the manifest gives for it, so a truncated
//   transfer that a proxy reported as success does not reach the cache.
//
//   A transient failure is retried with a doubling backoff, and a permanent one is not.
//
// Reading is somebody else's job. This produces a directory, and the generated tables
// read it.
//
// Foundation and nothing else. URLSession does the transfer - and on Linux and Windows it
// lives in FoundationNetworking, which is the one import here that has to be conditional.
// JSONSerialization reads the manifest, so unlike some of the other runtimes this one needs
// no JSON parser of its own. MD5 is written out below rather than taken from a crypto
// package, the same judgement the Rust updater makes about its own: one hash is cheaper to
// carry than a dependency the reader does not otherwise need.

import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public enum TabbitUpdater {

    /// What an update is allowed to do. Every value has a working default.
    public struct Options {

        /// The binary exporter writes manifest-binary.json; the JSON exporter writes
        /// manifest-json.json.
        public var manifestFileName: String = "manifest-binary.json"

        /// The first attempt is included, so three is two retries.
        public var maxAttempts: Int = 3

        /// Waited before the second attempt, in seconds. Doubled for each attempt after it.
        public var retryDelay: TimeInterval = 0.5

        public var requestTimeout: TimeInterval = 30

        public var verifyHash: Bool = true

        /// Called with one line of progress, when set.
        public var log: ((String) -> Void)? = nil

        public init() {}
    }

    /// What an update did.
    public struct Result {

        /// The directory holding the data. Hand it to the generated accessor's readAll. Set
        /// even on failure, because the previous data is still there and still readable -
        /// which is the point of failing the way this does.
        public let localPath: String

        public var succeeded: Bool = false
        public var error: String? = nil
        public var upToDate: Bool = false
        public var downloadedCount: Int = 0
        public var downloadedBytes: Int = 0
        public var deletedCount: Int = 0

        init(localPath: String) { self.localPath = localPath }
    }

    /// One file of the manifest, and the hash to check it by.
    public struct ManifestEntry {
        public let name: String
        public let size: Int64
        public let hash: String
    }

    /// What one request came back with, filled in by the transfer's completion.
    ///
    /// `@unchecked Sendable` because the safety is the semaphore's rather than the type's:
    /// the completion writes these three and signals, and the caller reads them only after
    /// waiting. A struct would not do - the closure would be mutating captured `var`s, which
    /// Swift 6 refuses.
    private final class Answer: @unchecked Sendable {
        var payload: Data? = nil
        var response: URLResponse? = nil
        var failure: Error? = nil
    }

    /// A failure the same request might survive a moment later.
    private struct TransientError: Error { let message: String }

    /// A failure retrying cannot help.
    private struct PermanentError: Error { let message: String }

    private static func describe(_ error: Error) -> String {
        switch error {
        case let transient as TransientError: return transient.message
        case let permanent as PermanentError: return permanent.message
        case let tcb as TcbError: return tcb.message
        default: return "\(error)"
        }
    }

    /// Brings `cacheDirectory` up to date with the data served under `baseUrl`.
    ///
    /// Does not throw. Everything that can go wrong here - the network, the disk, a file
    /// that arrived corrupt - is a condition the caller has to handle rather than a defect,
    /// and a patcher that throws into a service's start-up is one that gets wrapped in a
    /// bare `try?` that swallows the reason.
    public static func update(
        _ baseUrl: String,
        cacheDirectory: String,
        options: Options = Options()
    ) -> Result {
        var result = Result(localPath: cacheDirectory)
        let files = FileManager.default

        do {
            let manifestData = try download(
                joinUrl(baseUrl, options.manifestFileName), options)

            guard let manifestText = String(data: manifestData, encoding: .utf8) else {
                throw PermanentError(message: "the manifest is not UTF-8 text.")
            }

            let remote = try parseManifest(manifestText)
            let local = readLocalManifest(
                localPath(cacheDirectory, options.manifestFileName))

            var byName = [String: ManifestEntry]()
            for entry in local { byName[entry.name] = entry }

            let wanted = remote.filter { entry in
                guard let previous = byName[entry.name] else { return true }

                // The file's presence is checked as well as the manifest's word for it: a
                // cache somebody cleaned out by hand would otherwise never be refilled.
                return previous.hash != entry.hash
                    || !files.fileExists(atPath: localPath(cacheDirectory, entry.name))
            }

            let served = Set(remote.map { $0.name })
            let gone = local.map { $0.name }.filter { !served.contains($0) }

            if wanted.isEmpty && gone.isEmpty {
                options.log?("tabbit: already up to date.")

                result.succeeded = true
                result.upToDate = true
                return result
            }

            options.log?("tabbit: \(wanted.count) file(s) to fetch, \(gone.count) to remove.")

            // Everything lands here first. Nothing the caller can read is touched until the
            // last file has arrived and been checked.
            let staging = localPath(cacheDirectory, ".staging")

            try files.createDirectory(
                atPath: cacheDirectory, withIntermediateDirectories: true)
            deleteRecursively(staging)
            try files.createDirectory(atPath: staging, withIntermediateDirectories: true)

            for entry in wanted {
                let payload = try download(joinUrl(baseUrl, entry.name), options)

                if options.verifyHash && !entry.hash.isEmpty {
                    let actual = md5Hex([UInt8](payload))

                    if actual.lowercased() != entry.hash.lowercased() {
                        throw PermanentError(message:
                            "'\(entry.name)' arrived with hash \(actual), and the manifest "
                            + "says \(entry.hash). Nothing was replaced.")
                    }
                }

                let staged = localPath(staging, entry.name)

                try files.createDirectory(
                    atPath: (staged as NSString).deletingLastPathComponent,
                    withIntermediateDirectories: true)

                try payload.write(to: URL(fileURLWithPath: staged))

                result.downloadedBytes += payload.count
            }

            // From here on the update is applied. Nothing below reaches the network.
            for name in gone {
                let target = localPath(cacheDirectory, name)

                if files.fileExists(atPath: target) {
                    try files.removeItem(atPath: target)
                }

                result.deletedCount += 1
            }

            for entry in wanted {
                let target = localPath(cacheDirectory, entry.name)

                try files.createDirectory(
                    atPath: (target as NSString).deletingLastPathComponent,
                    withIntermediateDirectories: true)

                // Replaced rather than moved onto: Foundation refuses a move whose
                // destination exists, and on the second update every destination does.
                if files.fileExists(atPath: target) {
                    try files.removeItem(atPath: target)
                }

                try files.moveItem(atPath: localPath(staging, entry.name), toPath: target)

                result.downloadedCount += 1
            }

            // Last, and that ordering is the recovery story: a run killed before this point
            // leaves a manifest describing the data that is still on disk, so the next run
            // fetches the same files again rather than believing it has them.
            try manifestData.write(
                to: URL(fileURLWithPath: localPath(cacheDirectory, options.manifestFileName)))

            deleteRecursively(staging)

            options.log?(
                "tabbit: updated. \(result.downloadedCount) fetched, "
                + "\(result.deletedCount) removed.")

            result.succeeded = true
            return result
        } catch {
            // The previous data is untouched, so the caller can carry on with it.
            result.error = describe(error)

            options.log?("tabbit: update failed: \(result.error ?? "")")
            return result
        }
    }

    /// Reads the entries out of a manifest's JSON.
    public static func parseManifest(_ text: String) throws -> [ManifestEntry] {
        guard let data = text.data(using: .utf8) else {
            throw PermanentError(message: "the manifest is not UTF-8 text.")
        }

        let parsed = try JSONSerialization.jsonObject(with: data)

        guard let manifest = parsed as? [String: Any] else {
            throw PermanentError(message: "the manifest is not an object.")
        }

        guard let items = manifest["Items"] as? [Any] else {
            throw PermanentError(message: "the manifest has no Items array.")
        }

        var entries = [ManifestEntry]()

        for item in items {
            guard let fields = item as? [String: Any] else { continue }
            guard let name = fields["Name"] as? String, !name.isEmpty else { continue }

            let size = (fields["Size"] as? NSNumber)?.int64Value ?? 0
            let hash = fields["Hash"] as? String ?? ""

            entries.append(ManifestEntry(name: name, size: size, hash: hash))
        }

        return entries
    }

    // ------------------------------------------------------------------ transfer

    /// Fetches one URL, retrying what is worth retrying.
    private static func download(_ url: String, _ options: Options) throws -> Data {
        var delay = max(0, options.retryDelay)
        let attempts = max(1, options.maxAttempts)
        var attempt = 1

        while true {
            do {
                return try fetch(url, options)
            } catch let error as TransientError {
                if attempt >= attempts { throw error }

                options.log?(String(
                    format: "tabbit: %@ Retrying in %.1fs (%d of %d).",
                    error.message, delay, attempt, attempts))

                Thread.sleep(forTimeInterval: delay)

                // Doubling rather than a fixed wait: a server refusing because it is
                // overloaded is not helped by every client coming back at the same interval.
                delay *= 2
                attempt += 1
            }
        }
    }

    /// One request, waited for.
    ///
    /// URLSession has no synchronous form, and this call is synchronous because everything
    /// around it is: a data update happens at start-up or at a loading screen, where the
    /// caller's next line depends on it. The completion runs on the session's own queue
    /// rather than the caller's, so waiting here does not deadlock the thread that signals.
    private static func fetch(_ url: String, _ options: Options) throws -> Data {
        guard let target = URL(string: url) else {
            throw PermanentError(message: "'\(url)' is not a URL.")
        }

        var request = URLRequest(url: target)
        request.timeoutInterval = options.requestTimeout
        request.httpMethod = "GET"

        // A box rather than three captured locals. Swift 6 makes mutating a captured `var`
        // from a concurrently-executing closure an error, and it is right to: what makes
        // this safe is that nothing reads the fields until the semaphore has been signalled,
        // which the compiler cannot see. Saying so with `@unchecked Sendable` puts the claim
        // where a reader will find it.
        let answer = Answer()
        let finished = DispatchSemaphore(value: 0)

        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            answer.payload = data
            answer.response = response
            answer.failure = error
            finished.signal()
        }

        task.resume()
        finished.wait()

        let payload = answer.payload
        let response = answer.response

        if let error = answer.failure {
            // The request never got an answer - DNS, a refused connection, a timeout.
            throw TransientError(
                message: "'\(url)' could not be reached: \(error.localizedDescription).")
        }

        guard let status = (response as? HTTPURLResponse)?.statusCode else {
            throw TransientError(message: "'\(url)' gave no HTTP response.")
        }

        if status >= 200 && status <= 299 {
            return payload ?? Data()
        }

        let message = "'\(url)' answered \(status)."

        // 408 and 429 are the server asking for another attempt, and 5xx is it failing on
        // its own account. A 404 is an answer: retrying it costs three round trips to hear
        // the same thing.
        if status == 408 || status == 429 || (status >= 500 && status <= 599) {
            throw TransientError(message: message)
        }

        throw PermanentError(message: message)
    }

    // --------------------------------------------------------------------- disk

    /// Reads the cached manifest.
    ///
    /// A missing or unreadable one is an empty manifest, which makes the next update fetch
    /// everything - the safe direction to be wrong in.
    private static func readLocalManifest(_ file: String) -> [ManifestEntry] {
        guard let data = FileManager.default.contents(atPath: file),
              let text = String(data: data, encoding: .utf8) else { return [] }

        return (try? parseManifest(text)) ?? []
    }

    /// A manifest name resolved under a directory, with its forward slashes honoured.
    private static func localPath(_ directory: String, _ name: String) -> String {
        var resolved = URL(fileURLWithPath: directory)

        for part in name.split(separator: "/") where !part.isEmpty {
            resolved = resolved.appendingPathComponent(String(part))
        }

        return resolved.path
    }

    private static func deleteRecursively(_ directory: String) {
        // Failure is not reported: this runs to clear a staging directory that may not be
        // there, and a directory that cannot be removed shows up as the next step failing
        // with a reason of its own.
        try? FileManager.default.removeItem(atPath: directory)
    }

    /// Joins a base URL and a file name.
    ///
    /// Not a path join, which on Windows produces a backslash and a URL no server will
    /// answer.
    private static func joinUrl(_ baseUrl: String, _ name: String) -> String {
        var base = baseUrl

        while base.hasSuffix("/") { base.removeLast() }

        return base + "/" + name.replacingOccurrences(of: "\\", with: "/")
    }

    // ---------------------------------------------------------------------- MD5

    /// The MD5 of some bytes, in the lower-case hex the manifest carries.
    ///
    /// RFC 1321, written out rather than taken from CryptoKit or swift-crypto. What the
    /// hash is for is catching a transfer that arrived short, not resisting an adversary -
    /// so the reason to reach for a crypto package is not present, and the reason not to
    /// is: this file would then require one on every platform, for a hundred lines of
    /// arithmetic. The reader's own MAC is the other way round, and
    /// spec/tcb-mac-and-signature.md says why.
    ///
    /// Every operation wraps. Swift traps on overflow, and MD5 is nothing but wrapping
    /// addition and rotation.
    public static func md5Hex(_ data: [UInt8]) -> String {
        var a0: UInt32 = 0x6745_2301
        var b0: UInt32 = 0xefcd_ab89
        var c0: UInt32 = 0x98ba_dcfe
        var d0: UInt32 = 0x1032_5476

        // The message, padded: a 0x80 byte, zeros to eight short of a block, then the
        // length in bits as a little-endian 64.
        var message = data
        message.append(0x80)

        while message.count % 64 != 56 { message.append(0) }

        let bits = UInt64(data.count) &* 8
        for shift in stride(from: 0, to: 64, by: 8) {
            message.append(UInt8(truncatingIfNeeded: bits &>> UInt64(shift)))
        }

        var block = [UInt32](repeating: 0, count: 16)

        for start in stride(from: 0, to: message.count, by: 64) {
            for word in 0 ..< 16 {
                let at = start + word * 4

                block[word] = UInt32(message[at])
                    | (UInt32(message[at + 1]) &<< 8)
                    | (UInt32(message[at + 2]) &<< 16)
                    | (UInt32(message[at + 3]) &<< 24)
            }

            var a = a0
            var b = b0
            var c = c0
            var d = d0

            for round in 0 ..< 64 {
                var f: UInt32
                var g: Int

                switch round / 16 {
                case 0:
                    f = (b & c) | (~b & d)
                    g = round
                case 1:
                    f = (d & b) | (~d & c)
                    g = (5 * round + 1) % 16
                case 2:
                    f = b ^ c ^ d
                    g = (3 * round + 5) % 16
                default:
                    f = c ^ (b | ~d)
                    g = (7 * round) % 16
                }

                f = f &+ a &+ md5Constants[round] &+ block[g]

                a = d
                d = c
                c = b
                b = b &+ rotate(f, md5Shifts[round])
            }

            a0 = a0 &+ a
            b0 = b0 &+ b
            c0 = c0 &+ c
            d0 = d0 &+ d
        }

        var hex = ""
        hex.reserveCapacity(32)

        for word in [a0, b0, c0, d0] {
            for shift in stride(from: 0, to: 32, by: 8) {
                let byte = UInt8(truncatingIfNeeded: word &>> UInt32(shift))
                hex.append(md5Hexits[Int(byte >> 4)])
                hex.append(md5Hexits[Int(byte & 0x0F)])
            }
        }

        return hex
    }

    private static func rotate(_ value: UInt32, _ by: UInt32) -> UInt32 {
        (value &<< by) | (value &>> (32 &- by))
    }

    private static let md5Hexits = Array("0123456789abcdef")

    /// The per-round shift amounts of RFC 1321.
    private static let md5Shifts: [UInt32] = [
        7, 12, 17, 22, 7, 12, 17, 22, 7, 12, 17, 22, 7, 12, 17, 22,
        5, 9, 14, 20, 5, 9, 14, 20, 5, 9, 14, 20, 5, 9, 14, 20,
        4, 11, 16, 23, 4, 11, 16, 23, 4, 11, 16, 23, 4, 11, 16, 23,
        6, 10, 15, 21, 6, 10, 15, 21, 6, 10, 15, 21, 6, 10, 15, 21,
    ]

    /// The sine-derived constants of RFC 1321.
    private static let md5Constants: [UInt32] = [
        0xd76a_a478, 0xe8c7_b756, 0x2420_70db, 0xc1bd_ceee,
        0xf57c_0faf, 0x4787_c62a, 0xa830_4613, 0xfd46_9501,
        0x6980_98d8, 0x8b44_f7af, 0xffff_5bb1, 0x895c_d7be,
        0x6b90_1122, 0xfd98_7193, 0xa679_438e, 0x49b4_0821,
        0xf61e_2562, 0xc040_b340, 0x265e_5a51, 0xe9b6_c7aa,
        0xd62f_105d, 0x0244_1453, 0xd8a1_e681, 0xe7d3_fbc8,
        0x21e1_cde6, 0xc337_07d6, 0xf4d5_0d87, 0x455a_14ed,
        0xa9e3_e905, 0xfcef_a3f8, 0x676f_02d9, 0x8d2a_4c8a,
        0xfffa_3942, 0x8771_f681, 0x6d9d_6122, 0xfde5_380c,
        0xa4be_ea44, 0x4bde_cfa9, 0xf6bb_4b60, 0xbebf_bc70,
        0x289b_7ec6, 0xeaa1_27fa, 0xd4ef_3085, 0x0488_1d05,
        0xd9d4_d039, 0xe6db_99e5, 0x1fa2_7cf8, 0xc4ac_5665,
        0xf429_2244, 0x432a_ff97, 0xab94_23a7, 0xfc93_a039,
        0x655b_59c3, 0x8f0c_cc92, 0xffef_f47d, 0x8584_5dd1,
        0x6fa8_7e4f, 0xfe2c_e6e0, 0xa301_4314, 0x4e08_11a1,
        0xf753_7e82, 0xbd3a_f235, 0x2ad7_d2bb, 0xeb86_d391,
    ]
}

File: lib/lua/tabbit/updater.lua
-- Tabbit's data updater.
--
-- Copied in beside the generated accessor so the emitted code needs nothing installed.
-- Edit it in the Tabbit repository.
--
-- Brings a local copy of the exported data up to date with a copy served over HTTP - a
-- CDN, a bucket, a patch server - so a running program can take new data without being
-- redeployed. Reads nothing but the manifest, so it knows nothing about the schema and
-- never has to change when one does.
--
-- Three properties, because a patcher that fails badly is worse than one that does not
-- exist: nothing is replaced until everything has arrived and been checked; every file
-- is checked against the hash the manifest gives for it; a transient failure is retried
-- with a doubling backoff and a permanent one is not.
--
-- Two things are not this file's:
--
--   HTTP. Lua's standard library has none, and the program embedding this - a game
--   client above all - has its own stack already; requiring a second one would be the
--   cost with no benefit. `options.fetch` is how the transfer comes in: a function
--   taking a URL and returning the body, or nil, a message, and whether the failure is
--   worth retrying. spec/lua-language-support.md.
--
--   The hash, the directories and the waiting, which come from tabbit.native - the MD5
--   is a byte loop Lua cannot afford, and mkdir and a sleep are simply not in os.

local _prefix = (...):match("^(.-)[^%.]*$")

local function native_module()
  local ok, native = pcall(require, _prefix .. "native")

  if ok then
    return native
  end

  error("tcb: the updater needs the tabbit.native module, and it is not loadable - " ..
    "compile tabbit/native/tabbit_native.c into your host or build it as a Lua module", 0)
end

local updater = {}

-- The entries of a manifest's JSON: name, size, hash per file. A narrow parser rather
-- than a JSON library - the manifest is the exporter's, its shape does not vary, and
-- neither supported runtime ships a parser to borrow.
local function parse_manifest(text)
  local at = 1

  local function skip_space()
    at = string.find(text, "[^ \t\r\n]", at) or #text + 1
  end

  local function fail_at(what)
    error(("the manifest is not readable: %s at byte %d"):format(what, at), 0)
  end

  local parse_value

  local function parse_string()
    if text:sub(at, at) ~= "\"" then
      fail_at("expected a string")
    end

    local out = {}
    at = at + 1

    while true do
      local c = text:sub(at, at)

      if c == "" then
        fail_at("unterminated string")
      elseif c == "\"" then
        at = at + 1
        return table.concat(out)
      elseif c == "\\" then
        local escape = text:sub(at + 1, at + 1)
        local plain = { ["\""] = "\"", ["\\"] = "\\", ["/"] = "/", b = "\b",
                        f = "\f", n = "\n", r = "\r", t = "\t" }

        if plain[escape] ~= nil then
          out[#out + 1] = plain[escape]
          at = at + 2
        elseif escape == "u" then
          -- The manifest's names are file names; a code point above 7F is carried
          -- through as the replacement the writer used, which for this file is none.
          local hex = text:sub(at + 2, at + 5)
          local code = tonumber(hex, 16) or fail_at("a broken \\u escape")

          if code < 0x80 then
            out[#out + 1] = string.char(code)
          else
            out[#out + 1] = "?"
          end

          at = at + 6
        else
          fail_at("an unknown escape")
        end
      else
        out[#out + 1] = c
        at = at + 1
      end
    end
  end

  local function parse_number()
    local from = at
    at = string.find(text, "[^-+0-9eE.]", at) or #text + 1
    return tonumber(text:sub(from, at - 1)) or fail_at("a broken number")
  end

  parse_value = function()
    skip_space()
    local c = text:sub(at, at)

    if c == "\"" then
      return parse_string()
    elseif c == "{" then
      local out = {}
      at = at + 1
      skip_space()

      if text:sub(at, at) == "}" then
        at = at + 1
        return out
      end

      while true do
        skip_space()
        local key = parse_string()
        skip_space()

        if text:sub(at, at) ~= ":" then
          fail_at("expected a colon")
        end

        at = at + 1
        out[key] = parse_value()
        skip_space()

        local nextc = text:sub(at, at)
        at = at + 1

        if nextc == "}" then
          return out
        elseif nextc ~= "," then
          fail_at("expected , or }")
        end
      end
    elseif c == "[" then
      local out = {}
      at = at + 1
      skip_space()

      if text:sub(at, at) == "]" then
        at = at + 1
        return out
      end

      while true do
        out[#out + 1] = parse_value()
        skip_space()

        local nextc = text:sub(at, at)
        at = at + 1

        if nextc == "]" then
          return out
        elseif nextc ~= "," then
          fail_at("expected , or ]")
        end
      end
    elseif text:sub(at, at + 3) == "true" then
      at = at + 4
      return true
    elseif text:sub(at, at + 4) == "false" then
      at = at + 5
      return false
    elseif text:sub(at, at + 3) == "null" then
      at = at + 4
      return nil
    else
      return parse_number()
    end
  end

  local document = parse_value()
  local items = type(document) == "table" and document.Items or nil

  if type(items) ~= "table" then
    error("the manifest has no Items array", 0)
  end

  local entries = {}

  for i = 1, #items do
    local item = items[i]

    if type(item) == "table" and item.Name ~= nil and item.Name ~= "" then
      entries[#entries + 1] = {
        name = item.Name,
        size = item.Size or 0,
        hash = item.Hash or "",
      }
    end
  end

  return entries
end

updater.parseManifest = parse_manifest

-- Joins a base URL and a file name. Not a path join, which on Windows would produce a
-- backslash and a URL no server will answer.
local function join_url(base_url, name)
  return (base_url:gsub("/+$", "")) .. "/" .. (name:gsub("\\", "/"))
end

local function join_path(directory, name)
  return directory .. "/" .. (name:gsub("\\", "/"))
end

-- Creates a path's directories, one level at a time - native.mkdir is one mkdir(2), and
-- a level that already exists answers an error this ignores.
local function make_directories(native, path)
  local built = nil

  for piece in path:gmatch("[^/\\]+") do
    built = built and (built .. "/" .. piece) or piece
    native.mkdir(built)
  end
end

local function read_file(path)
  local handle = io.open(path, "rb")

  if handle == nil then
    return nil
  end

  local data = handle:read("*a")
  handle:close()

  return data
end

local function write_file(path, data)
  local handle, err = io.open(path, "wb")

  if handle == nil then
    error(("cannot write %s: %s"):format(path, err or "unknown error"), 0)
  end

  handle:write(data)
  handle:close()
end

-- The cached manifest. A missing or unreadable one is an empty manifest, which makes
-- the next update fetch everything - the safe direction to be wrong in.
local function read_local_manifest(path)
  local text = read_file(path)

  if text == nil then
    return {}
  end

  local ok, entries = pcall(parse_manifest, text)

  return ok and entries or {}
end

-- Fetches one URL through options.fetch, retrying what the fetch says is worth
-- retrying, with a doubling backoff.
local function download(native, url, options, log)
  local delay = options.retryDelay or 0.5
  local attempts = math.max(1, options.maxAttempts or 3)

  for attempt = 1, attempts do
    local body, message, transient = options.fetch(url)

    if body ~= nil then
      return body
    end

    if not transient or attempt >= attempts then
      error(message or (url .. " could not be fetched"), 0)
    end

    log(("tabbit: %s Retrying in %.1fs (%d of %d)."):format(
      message or "fetch failed.", delay, attempt, attempts))

    native.sleepMs(math.floor(delay * 1000))

    -- Doubling rather than a fixed wait: a server refusing because it is overloaded is
    -- not helped by every client coming back at the same interval.
    delay = delay * 2
  end
end

-- Brings `cacheDirectory` up to date with the data served under `baseUrl`.
--
-- Does not raise. Everything that can go wrong here - the network, the disk, a file
-- that arrived corrupt - is a condition the caller has to handle rather than a defect.
-- The result says what happened; `localPath` is set even on failure, because the
-- previous data is still there and still readable, which is the point of failing the
-- way this does.
--
-- `options.fetch` is required - see the note at the top of this file. Everything else
-- has a working default: manifestFileName, maxAttempts, retryDelay (seconds),
-- verifyHash, log.
function updater.update(base_url, cache_directory, options)
  options = options or {}

  local result = {
    succeeded = false,
    error = nil,
    upToDate = false,
    downloadedCount = 0,
    downloadedBytes = 0,
    deletedCount = 0,
    localPath = cache_directory,
  }

  local function log(message)
    if options.log ~= nil then
      options.log(message)
    end
  end

  local ok, failure = pcall(function()
    if type(options.fetch) ~= "function" then
      error("options.fetch is required: a function taking a URL and returning the " ..
        "body, or nil, a message, and whether the failure is transient", 0)
    end

    local native = native_module()
    local manifest_name = options.manifestFileName or "manifest-binary.json"
    local verify = options.verifyHash ~= false

    local manifest_text = download(native, join_url(base_url, manifest_name), options, log)
    local remote = parse_manifest(manifest_text)
    local local_entries = read_local_manifest(join_path(cache_directory, manifest_name))

    local by_name = {}

    for i = 1, #local_entries do
      by_name[local_entries[i].name] = local_entries[i]
    end

    local wanted = {}

    for i = 1, #remote do
      local entry = remote[i]
      local previous = by_name[entry.name]

      -- The file's presence is checked as well as the manifest's word for it: a cache
      -- somebody cleaned out by hand would otherwise never be refilled.
      local current = previous ~= nil
        and previous.hash == entry.hash
        and read_file(join_path(cache_directory, entry.name)) ~= nil

      if not current then
        wanted[#wanted + 1] = entry
      end
    end

    local served = {}

    for i = 1, #remote do
      served[remote[i].name] = true
    end

    local gone = {}

    for i = 1, #local_entries do
      if not served[local_entries[i].name] then
        gone[#gone + 1] = local_entries[i].name
      end
    end

    if #wanted == 0 and #gone == 0 then
      log("tabbit: already up to date.")

      result.succeeded = true
      result.upToDate = true
      return
    end

    log(("tabbit: %d file(s) to fetch, %d to remove."):format(#wanted, #gone))

    -- Everything lands here first. Nothing the caller can read is touched until the
    -- last file has arrived and been checked. Stale files a killed run left behind are
    -- harmless: only this run's names are moved out, and a name this run downloads is
    -- written over.
    local staging = join_path(cache_directory, ".staging")

    make_directories(native, cache_directory)
    make_directories(native, staging)

    for i = 1, #wanted do
      local entry = wanted[i]
      local data = download(native, join_url(base_url, entry.name), options, log)

      if verify and entry.hash ~= "" then
        local actual = native.md5hex(data)

        if actual:lower() ~= entry.hash:lower() then
          error(("%s arrived with hash %s, and the manifest says %s. Nothing was " ..
            "replaced."):format(entry.name, actual, entry.hash), 0)
        end
      end

      local staged = join_path(staging, entry.name)

      make_directories(native, staged:match("^(.*)[/\\][^/\\]+$") or "")
      write_file(staged, data)

      result.downloadedBytes = result.downloadedBytes + #data
    end

    -- From here on the update is applied. Nothing below reaches the network.
    for i = 1, #gone do
      os.remove(join_path(cache_directory, gone[i]))
      result.deletedCount = result.deletedCount + 1
    end

    for i = 1, #wanted do
      local entry = wanted[i]
      local target = join_path(cache_directory, entry.name)

      make_directories(native, target:match("^(.*)[/\\][^/\\]+$") or "")

      -- Removed first: os.rename does not replace an existing file on Windows.
      os.remove(target)
      local moved, err = os.rename(join_path(staging, entry.name), target)

      if not moved then
        error(("cannot move %s into place: %s"):format(entry.name, err or "unknown"), 0)
      end

      result.downloadedCount = result.downloadedCount + 1
    end

    -- Last, and that ordering is the recovery story: a run killed before this point
    -- leaves a manifest describing the data that is still on disk, so the next run
    -- fetches the same files again rather than believing it has them.
    write_file(join_path(cache_directory, manifest_name), manifest_text)

    -- os.remove takes an empty directory where the platform's remove() does; where it
    -- does not, an empty .staging stays behind and costs nothing.
    os.remove(staging)

    log(("tabbit: updated. %d fetched, %d removed."):format(
      result.downloadedCount, result.deletedCount))

    result.succeeded = true
  end)

  if not ok then
    -- The previous data is untouched, so the caller can carry on with it.
    result.error = tostring(failure)
    result.succeeded = false

    log("tabbit: update failed: " .. result.error)
  end

  return result
end

return updater
